### PR TITLE
implement EIP-234: add blockHash param for eth_getLogs

### LINKED ISSUE #457: The method `eth_getLogs` does not support `blockHash` param
When making a rpc query to `https://rpc.ankr.com/xdc`

```
{
    "jsonrpc": "2.0",
    "method": "eth_getLogs",
    "params": [
        {
            "blockHash": "0xdc480998d4335c61fe5d32802b455a6fc5fa8c08e1b55ebd4c9229adde15e99c"
        }
    ],
    "id": 1
}
```

it would return the log for the latest block instead of the blockHash provided in the param.
if this is expected result then https://www.ankr.com/docs/rpc-service/chains/chains-api/xdc/#parameters-27 should be updated.

### PR DESCRIPTION
**Pull request by gzliudan**
# Proposed changes

This PR fix #457 by support parameter `blockHash` in the input of method `eth_getLogs`.

**Notice: this function is removed by PR #131.**

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [X] Geth
- [ ] Smart Contract
- [ ] External components
- [ ] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [X] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

## Test

---

**request:**

```bash
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 6003,
  "method": "eth_getLogs",
  "params": [
    {
      "blockHash": "'"${HASH}"'"
    }
  ]
}' | jq
```

**response:**

```json
{
  "jsonrpc": "2.0",
  "id": 6003,
  "result": [
    {
      "address": "0x0000000000000000000000000000000000000089",
      "topics": [
        "0x62855fa22e051687c32ac285857751f6d3f2c100c72756d8d30cb7ecb1f64f54"
      ],
      "data": "0x00000000000000000000000096509a56f0243b10a2706391b95af449712da69900000000000000000000000000000000000000000000000000000000000005dcce1d2eca75e10aeea5edf82ac30b24d226881a4e6eae4b89120b73978862aa9c",
      "blockNumber": "0x5de",
      "transactionHash": "0x7f58b867d33756af9a5d42eba7955f943a99a05516da672a77210300152b08ce",
      "transactionIndex": "0x0",
      "blockHash": "0xd56a4b9835ad8067cf87e29d1a12437369fb4a4bd1a5b5cf5d538e7fd34806e4",
      "logIndex": "0x0",
      "removed": false
    },
    {
      "address": "0x0000000000000000000000000000000000000089",
      "topics": [
        "0x62855fa22e051687c32ac285857751f6d3f2c100c72756d8d30cb7ecb1f64f54"
      ],
      "data": "0x0000000000000000000000002526fdebae27162e45da84e5e2f00abaa1a1cc8c00000000000000000000000000000000000000000000000000000000000005dcce1d2eca75e10aeea5edf82ac30b24d226881a4e6eae4b89120b73978862aa9c",
      "blockNumber": "0x5de",
      "transactionHash": "0x729eb00a6f54f1fb7aaf6f012b6b57b263f820a2b3b6781b42a6c84314171a8c",
      "transactionIndex": "0x1",
      "blockHash": "0xd56a4b9835ad8067cf87e29d1a12437369fb4a4bd1a5b5cf5d538e7fd34806e4",
      "logIndex": "0x1",
      "removed": false
    }
  ]
}
```

---

**request:**

```bash
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
   "jsonrpc": "2.0",
   "id": 6003,
   "method": "eth_getLogs",
   "params": [
     {
       "fromBlock": "0x1",
       "toBlock": "0xf"
     },
     {
       "blockHash": "'"${HASH}"'"
     }
   ]
}' | jq
```

**response:**

```json
{
  "jsonrpc": "2.0",
  "id": 6003,
  "error": {
    "code": -32602,
    "message": "too many arguments, want at most 1"
  }
}
```

---

**request:**

```bash
curl -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc":"2.0",
  "id": 6002,
  "method":"eth_getLogs",
  "params":[
    {
      "fromBlock": "0x1",
      "toBlock": "0xff",
      "topics": [
        "0x000000000000000000000000a94f5374fce5edbc8e2a8697c15331677e6ebf0b"
      ]
    }
  ]
}' | jq
```

**response:**

```json
{
  "jsonrpc": "2.0",
  "id": 6002,
  "result": []
}
```

---

**request:**

```bash
curl -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc":"2.0",
  "id": 6002,
  "method":"eth_getLogs",
  "params":[
    {
      "fromBlock": "0x1",
      "toBlock": "0x10"
    }
  ]
}' | jq
```

**response**

```json
{
  "jsonrpc": "2.0",
  "id": 6002,
  "result": [
    {
      "address": "0x0000000000000000000000000000000000000089",
      "topics": [
        "0x62855fa22e051687c32ac285857751f6d3f2c100c72756d8d30cb7ecb1f64f54"
      ],
      "data": "0x0000000000000000000000002526fdebae27162e45da84e5e2f00abaa1a1cc8c000000000000000000000000000000000000000000000000000000000000000f4027ffb371d54b5c072b618a00724fda5907ff324e557f3be50102502597a60e",
      "blockNumber": "0x10",
      "transactionHash": "0xdea69962667164bfa0a76d2678c0dc9f9a3d7226a8c59e9d6a4ba3796f219067",
      "transactionIndex": "0x0",
      "blockHash": "0x07745cd6967cd94bd401bbdfe4b44f6b3953ff11e1a59bc284dc466647f044db",
      "logIndex": "0x0",
      "removed": false
    },
    {
      "address": "0x0000000000000000000000000000000000000089",
      "topics": [
        "0x62855fa22e051687c32ac285857751f6d3f2c100c72756d8d30cb7ecb1f64f54"
      ],
      "data": "0x00000000000000000000000096509a56f0243b10a2706391b95af449712da699000000000000000000000000000000000000000000000000000000000000000f4027ffb371d54b5c072b618a00724fda5907ff324e557f3be50102502597a60e",
      "blockNumber": "0x10",
      "transactionHash": "0x0ac0dca577b9dc3bd06e7f58dc55225178b8da24920ebfec6f17916268165ea9",
      "transactionIndex": "0x1",
      "blockHash": "0x07745cd6967cd94bd401bbdfe4b44f6b3953ff11e1a59bc284dc466647f044db",
      "logIndex": "0x1",
      "removed": false
    },
    {
      "address": "0x0000000000000000000000000000000000000089",
      "topics": [
        "0x62855fa22e051687c32ac285857751f6d3f2c100c72756d8d30cb7ecb1f64f54"
      ],
      "data": "0x00000000000000000000000077cb85ae0ae070dfc013ba1a5b3ee1ced4a059a7000000000000000000000000000000000000000000000000000000000000000f4027ffb371d54b5c072b618a00724fda5907ff324e557f3be50102502597a60e",
      "blockNumber": "0x10",
      "transactionHash": "0x148c22e50bd86b8109679b6e0c51be5dbd2813e8779c2fafcd25fb198e6183f9",
      "transactionIndex": "0x2",
      "blockHash": "0x07745cd6967cd94bd401bbdfe4b44f6b3953ff11e1a59bc284dc466647f044db",
      "logIndex": "0x2",
      "removed": false
    }
  ]
}
```

### CODE REVIEW

File: eth/filters/api.go [hunk 1]
```diff
@@ -327,15 +328,38 @@ func (api *PublicFilterAPI) NewFilter(crit FilterCriteria) (rpc.ID, error) {
...
+		fromBlock = int64(header.Number.Int64())
+		toBlock = fromBlock
+	} else {
+		// Convert the RPC block numbers into internal representations
+		if crit.FromBlock == nil {
+			fromBlock = int64(rpc.LatestBlockNumber)
+		}
+		if crit.ToBlock == nil {
+			toBlock = int64(rpc.LatestBlockNumber)
+		}
```
Comment (liam-lai) [inline eth/filters/api.go:358]: If ToBlock != nil, should we assign toBlock = crit.ToBlock.Int64()? Same as fromBlock

File: eth/filters/api.go [hunk 2]
```diff
@@ -327,15 +328,38 @@ func (api *PublicFilterAPI) NewFilter(crit FilterCriteria) (rpc.ID, error) {
 //
 // https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_getlogs
 func (api *PublicFilterAPI) GetLogs(ctx context.Context, crit FilterCriteria) ([]*types.Log, error) {
-	// Convert the RPC block numbers into internal representations
-	if crit.FromBlock == nil {
-		crit.FromBlock = big.NewInt(rpc.LatestBlockNumber.Int64())
+	if len(crit.Topics) > maxTopics {
```
Comment (liam-lai) [inline eth/filters/api.go:331]: what's this Topics about?

File: eth/filters/api.go (hunk 1 above)
Comment (gzliudan) [reply eth/filters/api.go:362]: yes

File: eth/filters/api.go (hunk 2 above)
Comment (gzliudan) [reply eth/filters/api.go:331]: The topics field can contain up to 4 topics. The first topic is required and will always contain the keccak 256 hash of the event signature. The other three topics are optional and typically used for indexing and provide a faster lookup time than using the data field described below.

**Review by liam-lai — APPROVED** [review]

**Pull request merged**